src/serlio/modifiers/RuleAttributes.cpp

```diff
@@ -214,16 +212,45 @@ bool RuleAttributeCmp::operator()(const RuleAttribute& lhs, const RuleAttribute&
 		return true;
 	};
 
-	auto firstDifferentGroupInA = [](const RuleAttribute& a, const RuleAttribute& b) {
-		assert(a.groups.size() == b.groups.size());
-		size_t i = 0;
-		while ((i < a.groups.size()) && (a.groups[i] == b.groups[i])) {
-			i++;
+	auto compareGroups = [](const RuleAttribute& a, const RuleAttribute& b) {
+		const size_t GroupSizeA = a.groups.size();
+		const size_t GroupSizeB = b.groups.size();
+
+		for (size_t i = 0; i < std::max(GroupSizeA, GroupSizeB); ++i) {
+			// a descendant of b
+			if (i >= GroupSizeA)
+				return false;
+
+			// b descendant of a
+			if (i >= GroupSizeB)
+				return true;
```

Comment:
Sadly we can't return early because we don't know if group length or an alphabetical difference in the group will decide which is sorted first.
Imagine two attributes a and b with multiple groups but the same group order
attr a groups: A>B>C
attr b groups: A>D>E>F
=> a before b

attr a groups: A>B
attr b groups: A
=> B before A